core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -486,12 +487,15 @@ private[kafka] class Processor(val id: Int,
   private def processCompletedReceives() {
     selector.completedReceives.asScala.foreach { receive =>
       try {
-        val channel = selector.channel(receive.source)
-        val session = RequestChannel.Session(new KafkaPrincipal(KafkaPrincipal.USER_TYPE, channel.principal.getName),
-          channel.socketAddress)
+        val openChannel = selector.channel(receive.source)
+        val session = {
+          val channel = if (openChannel != null) openChannel else selector.closedChannel(receive.source)
```

Comment:
Probably worth adding a comment noting that we should only call methods that are safe to call on a closed channel (or something to that effect).